cmd/kubeadm/app/cmd/alpha/certs.go

```diff
@@ -82,9 +129,111 @@ func newCmdCertsUtility(out io.Writer) *cobra.Command {
 	cmd.AddCommand(newCmdCertsRenewal(out))
 	cmd.AddCommand(newCmdCertsExpiration(out, constants.KubernetesDir))
 	cmd.AddCommand(NewCmdCertificateKey())
+	cmd.AddCommand(newCmdGenCSR())
+	return cmd
+}
+
+// genCSRConfig is the configuration required by the gencsr command
+type genCSRConfig struct {
+	kubeadmConfigPath string
+	certDir           string
+	kubeConfigDir     string
+	kubeadmConfig     *kubeadmapi.InitConfiguration
+}
+
+func newGenCSRConfig() *genCSRConfig {
+	return &genCSRConfig{
+		kubeConfigDir: kubeadmconstants.KubernetesDir,
+	}
+}
+
+func (o *genCSRConfig) addFlagSet(flagSet *pflag.FlagSet) {
+	options.AddConfigFlag(flagSet, &o.kubeadmConfigPath)
+	options.AddCertificateDirFlag(flagSet, &o.certDir)
+	options.AddKubeConfigDirFlag(flagSet, &o.kubeConfigDir)
+}
+
+// finalize merges command line flag values into kubeadm's config.
+// Loads Kubeadm config from a file (if present)
+// else use dynamically generated default config.
+// This configuration contains the DNS names and IP addresses which
+// are encoded in the control-plane CSRs.
+func (o *genCSRConfig) finalize() (err error) {
+	o.kubeadmConfig, err = configutil.LoadOrDefaultInitConfiguration(
```

Comment:
Is this command supposed to be invoked without a config file? If that's the case, then we may have to introduce the rest of the config command line flags here.

Reply:
@wallrj should correct me here, but i think if the user is setting custom SANs under ClusterConfiguration, then the new command should be aware of the --config.

Reply:
I'd need to add some of the other flags that are currently supported by `kubectl init phase certs all`

```
_output/bin/kubeadm init phase certs all --help
Generate all certificates

Usage:
  kubeadm init phase certs all [flags]

Flags:
      --apiserver-advertise-address string   The IP address the API Server will advertise it's listening on. If not set the default network interface will be used.
      --apiserver-cert-extra-sans strings    Optional extra Subject Alternative Names (SANs) to use for the API Server serving certificate. Can be both IP addresses and DNS names.
      --cert-dir string                      The path where to save and store the certificates. (default "/etc/kubernetes/pki")
      --config string                        Path to a kubeadm configuration file.
      --control-plane-endpoint string        Specify a stable IP address or DNS name for the control plane.
  -h, --help                                 help for all
      --kubernetes-version string            Choose a specific Kubernetes version for the control plane. (default "stable-1")
      --service-cidr string                  Use alternative range of IP address for service VIPs. (default "10.96.0.0/12")
      --service-dns-domain string            Use alternative domain for services, e.g. "myorg.internal". (default "cluster.local")


```

* `--api-server-cert-extra-sans`  https://github.com/kubernetes/kubernetes/blob/c5941e283fd7e37e953832d6743e7ea3b556ab7a/cmd/kubeadm/app/cmd/init.go#L248-L251 

Or I could require that the user supplies a kubeadm config file and puts all those settings in there.

Reply:
I think, that we can start simple and require a config file for now. If many users start complaining we can add these extra options.

Reply:
As discussed  in office hours meeting last week, I have not added the extra command line options.
The `--config` flag is optional. If  not supplied, the Kubeadm config will be dynamically generated based on local machine settings.